### PR TITLE
IGNITE-23673 SQL Calcite: Fix NVL/DECODE failure when used with param…

### PR DESCRIPTION
**Pull request by alex-plekhanov**
…eters of different data types

Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        register(SqlLibraryOperators.DECODE, IgniteSqlCallRewriteTable::decodeRewriter);
+    }
+
+    /** Registers rewriter for SQL operator. */
+    public void register(SqlOperator operator, BiFunction<SqlValidator, SqlCall, SqlCall> rewriter) {
+        map.put(operator, rewriter);
+    }
+
+    /** Rewrites SQL call. */
+    public SqlCall rewrite(SqlValidator validator, SqlCall call) {
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:59]: package-visible, not public??

File: modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/FunctionsTest.java [hunk 2]
```diff
@@ -301,4 +301,48 @@ public void testCastToBoolean() {
...
+        assertQuery("select nvl(?, ?)").withParams(1, "2").returns("1").check();
+        assertQuery("select nvl(?, ?)").withParams(1, 2d).returns(1d).check();
+        assertQuery("select nvl(?, ?)").withParams(null, 2).returns(2).check();
+        assertQuery("select nvl(?, ?)").withParams(null, "2").returns("2").check();
+        assertQuery("select nvl(?, ?)").withParams(null, null).returns(NULL_RESULT).check();
+    }
+
+    /** Tests DECODE with different parameters data types. */
+    @Test
+    public void testDecode() {
```
Comment (Vladsz83) [inline modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/FunctionsTest.java:328]: Suggestion: tests like `assertQuery("select decode(2, '1', 1, '2', 2::DOUBLE)").returns(2.0).check();`

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+    /** */
+    private IgniteSqlCallRewriteTable() {
+        register(SqlLibraryOperators.NVL, IgniteSqlCallRewriteTable::nvlRewriter);
+        register(SqlLibraryOperators.DECODE, IgniteSqlCallRewriteTable::decodeRewriter);
+    }
+
+    /** Registers rewriter for SQL operator. */
+    public void register(SqlOperator operator, BiFunction<SqlValidator, SqlCall, SqlCall> rewriter) {
+        map.put(operator, rewriter);
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:55]: Suggestion: any thread-safe implementation. We can't be sure that custom rewriters are registered from the same thread.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    /** Rewrites SQL call. */
+    SqlCall rewrite(SqlValidator validator, SqlCall call) {
+        BiFunction<SqlValidator, SqlCall, SqlCall> rewriter = map.get(call.getOperator());
+
+        return rewriter == null ? call : rewriter.apply(validator, call);
+    }
+
+    /** Rewrites NVL call to CASE WHEN call. */
+    public static SqlCall nvlRewriter(SqlValidator validator, SqlCall call) {
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:66]: private?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+            SqlNode elseExpr = operands.get(1);
+
+            return SqlCase.createSwitched(pos, null, whenList, thenList, elseExpr);
+        }
+        else
+            return call; // Operands count will be validated and exception will be thrown later.
+    }
+
+    /** Rewrites DECODE call to CASE WHEN call. */
+    public static SqlCall decodeRewriter(SqlValidator validator, SqlCall call) {
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:88]: private?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+        SqlParserPos pos = call.getParserPosition();
+
+        SqlNode op0 = operands.get(0);
+
+        SqlNodeList whenList = new SqlNodeList(pos);
+        SqlNodeList thenList = new SqlNodeList(pos);
+
+        for (int i = 1; i < operands.size() - 1; i += 2) {
+            whenList.add(SqlStdOperatorTable.IS_NOT_DISTINCT_FROM.createCall(pos, SqlNode.clone(op0), operands.get(i)));
```
Comment (Vladsz83) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:101]: Do we need `SqlNode.clone(op0)` here and above?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java (hunk 3 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:55]: Fixed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java (hunk 4 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java]: FIxed

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java (hunk 6 above)
Comment (alex-plekhanov) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteSqlCallRewriteTable.java:101]: Yes, we need to guarantee that oriented graph is a tree, or we can get some unexpected behaviour during recursive tree traversal.

**Pull request closed without merging**